redis-6.0.9/src/modules/systemtime.c

```diff
@@ -0,0 +1,246 @@
+#include <time.h>
+#include <sys/timerfd.h>
+#include <sys/epoll.h>
+#include <pthread.h>
+#include <unistd.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <string.h>
+#include "../redismodule.h"
+
+static int setup_io_multiplexer(RedisModuleCtx *ctx);
+static int setup_timer(RedisModuleCtx *ctx);
+static int start_sys_time_collector_thread(RedisModuleCtx *ctx);
+static void convert_time_to_human_string(time_t seconds, char* buffer);
+
+
+static int s_epoll_fd = 0;
+static int s_timer_fd = 0;
+static const int NANO_IN_SEC = 1000000000;
+static const int MILLI_IN_SEC = 1000;
+static long s_system_time_usec = 0;
+static double jiffi_in_seconds = 0.0;
+static pthread_mutex_t lock;
+static char s_formated_time_buffer[64];
+
+
+/*
+* DRIVENETS.GET.TIME
+* this is an implementation of an efficient get-system-time mechanism
+* the idea is to use an internal thread that will "collect" the time
+* passing into a static register. once a time sample is required, the
+* command api will sample _the register_ without calling an expensive
+* system call, or probing some hardware clock.
+*
+* the main idea is to move all the heavy lifting into the collector thread.
+* the thread executing the command will simply copy a - pre formatted - buffer
+* and send it via the redis internal ipc
+* profiling numbers are given for each build permutation in: profile_results
+*/
+int DrivenetsGetTime_RedisCommand(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
+	REDISMODULE_NOT_USED(argv);
+    REDISMODULE_NOT_USED(argc);
+	
+	// dont hold the lock for the ipc
+	char buffer[64] = {0};
+	pthread_mutex_lock(&lock);
+	memcpy(buffer, s_formated_time_buffer, 64);
```

Comment:
why the memcpy to buffer? why not
```
RedisModule_ReplyWithCString(ctx, s_formated_time_buffer);
```

Reply:
i didnt want to hold the mutex while doing the ipc.
i dont have a profile figure for this, so i dont know if this is meaningful in terms of performance (??)